File: 90. Subsets II/90. Subsets II.rb
# @param {Integer[]} nums
# @return {Integer[][]}
def subsets_with_dup(nums)
    nums.sort!
    result = []
    backtrack(nums, 0, [], result)
    result
end

def backtrack(nums, start, current, result)
    result << current.dup
    
    (start...nums.length).each do |i|
        next if i > start && nums[i] == nums[i-1]
        current << nums[i]
        backtrack(nums, i + 1, current, result)
        current.pop
    end
end

File: 90. Subsets II/90. Subsets II.js
function subsetsWithDup(nums) {
    const results = [];
    nums.sort((a, b) => a - b); // Sort to handle duplicates

    function backtrack(start, path) {
        results.push([...path]); // Add the current subset to the result
        for (let i = start; i < nums.length; i++) {
            if (i > start && nums[i] === nums[i - 1]) continue; // Skip duplicates
            path.push(nums[i]);
            backtrack(i + 1, path); // Recurse with the next starting index
            path.pop(); // Backtrack
        }
    }

    backtrack(0, []);
    return results;
}

// Example usage
console.log(subsetsWithDup([1, 2, 2])); // Output: [[], [1], [1, 2], [1, 2, 2], [2], [2, 2]]
console.log(subsetsWithDup([0]));        // Output: [[], [0]]